### ISSUE TITLE
@JiahuiYu . Hi, I'm sorry to bother you.But I still has some questions.

### ISSUE THREAD

**Issue by huangqianfirst** [JiahuiYu/generative_inpainting#180]
The change of loss.
the loss of D_net is:
g_loss, d_loss = self.gan_hinge_loss(pos_global, neg_global)
the complete loss of G loss is : 
losses['g_loss'] = a1 *  g_loss +a2 * losses['l1_loss']
Q1.What happens to this g_loss of D_net , it can drop to -19. Of cource ,this result is awful. what's the range of g_loss, d_loss?
Q2.**the a1, a2 that weights of complete losses['g_loss'] is 1 ?**
I'm looking forward to your reply. 

PS. I change loss below to make g_loss and l1 loss same order of magnitude.
d_loss = tf.reduce_mean(tf.nn.softplus(-pos)) + tf.reduce_mean(tf.nn.softplus(neg))
g_loss = tf.reduce_mean(tf.nn.softplus(-neg))
Below is my result. Still has artifacts.
![00000552_out_incp](https://user-images.githubusercontent.com/11473038/48749335-699e0700-ecb5-11e8-9c1f-f9adf084747e.png)
![00003101_out_incp](https://user-images.githubusercontent.com/11473038/48749338-6dca2480-ecb5-11e8-82e7-e97d293aa6b9.png)
![00002118_out_incp](https://user-images.githubusercontent.com/11473038/48749343-74f13280-ecb5-11e8-9674-cda0560ec6dd.png)

_Originally posted by @huangqianfirst in https://github.com/JiahuiYu/generative_inpainting/issues/172#issuecomment-438930420_

**Comment by JiahuiYu** [comment]
@huangqianfirst 

Hi, for Q1, I don't have an estimate of the range for g_loss. If GAN is converging, d_loss will have a increasing curve like from -10 to around -1.

For Q2, yes in deepfill v2 I use 1:1.

**Comment by huangqianfirst** [comment]
I use Titan Xp , 2 GPU, batchsize = 16. And I has been trained 192000 iters.The g loss is hard to go down.
Does this iters is enough or not ?

**Comment by JiahuiYu** [comment]
@huangqianfirst Usually I train for more than 800,000 iters for places2.

**Comment by huangqianfirst** [comment]
In  your mini-batchsize = 24?
And as my g_loss of D_net , it can drop to -19, i think it's weird. And hard to find result toward nice result. So I change like this , i think it' have no essential distinction. Right ?
d_loss = tf.reduce_mean(tf.nn.softplus(-pos)) + tf.reduce_mean(tf.nn.softplus(neg))
g_loss = tf.reduce_mean(tf.nn.softplus(-neg))
By the way ,do you remember the config , max_iters=1 in discriminator_training_callback, it's 1 or keep it old 5 ?

**Comment by JiahuiYu** [comment]
@huangqianfirst With spectral normalization, I use max_iters=1 for training. 

I use:
```
            pos_neg = self.build_gan_discriminator(batch_pos_neg, training=training, reuse=reuse)
            pos, neg = tf.split(pos_neg, 2)
            g_loss, d_loss = gan_hinge_loss(pos, neg)
```
where the gan_hinge_loss can be found in neuralgym dev branch.

**Comment by huangqianfirst** [comment]
Yes , i found it . And i use max_iters=1 for training. Since i found g_loss range greater(1 ~ -19) than l1_Loss( 1 ~0). So i thought it might beacuse D it's not trained enough.So I consider increasing the max_iters.

**Comment by JiahuiYu** [comment]
@huangqianfirst The loss does not mean anything in my view. What matters is the gradients of loss that are back-propagated to parameters. If d_loss converges, then GAN should be correct.

**Comment by huangqianfirst** [comment]
ok. I get it .Thanks for your rely. I'll spend more time to train.

**Comment by huangqianfirst** [comment]
@JiahuiYu . I still have questions.
You said : d_loss will have a increasing curve like from -10 to around -1. 
But d_loss is write like this:
hinge_pos = tf.reduce_mean(tf.nn.relu(1-pos))
hinge_neg = tf.reduce_mean(tf.nn.relu(1+neg
d_loss = tf.add(.5 * hinge_pos, .5 * hinge_neg)

It can't be smaller than zero. Isn't it ?

**Comment by JiahuiYu** [comment]
I mean the WGAN-GP case in this repo instead of SN-PatchGAN in DeepFill v2. Sorry for confusion. For SN-PatchGAN, I don’t remember what a typical curve is like.

**Comment by huangqianfirst** [comment]
That's all right. Thanks.

**Comment by huangqianfirst** [comment]
@JiahuiYu . Hi, sorry to bother  you again. I'm wonder if the blur texture in the result is caused by the G.
Here is my gated conv code : 
 x1 = tf.layers.conv2d(
        x, cnum, ksize, stride, dilation_rate=rate,
        activation=None, padding=padding, name=name + 'x1')
    x2 = tf.layers.conv2d(
        x, cnum, ksize, stride, dilation_rate=rate,
        activation=None, padding=padding, name=name + 'x2')
    if activation is not None:
        x = tf.nn.elu(x1) * tf.nn.sigmoid(x2)
    else:
        x = x1 * tf.nn.sigmoid(x2)
 I  try to modifie the last line to : x = x1 ，which you said，'the output is not gated convlution.'. But it make result worse. Could you give me some advice about how to make results have more texture detail.
The result blow is trained after 20 iters with 2 GPUs.
![wooden_out_498_202000](https://user-images.githubusercontent.com/11473038/49710258-4262a600-fc73-11e8-9601-f689fd6d3223.png)

**Comment by JiahuiYu** [comment]
Do you mean that you don’t have activation for discriminator? Please use same layer as in this code, I.e., leaky relu.

**Comment by huangqianfirst** [comment]
No, I use leaky relu. I mean i'm not use batchNormal before activation. I've test add batch normal make result worse. Could you please check my questions? Thanks.

**Comment by JiahuiYu** [comment]
You said you modified the last line to : x = x1. That is why I think you probably have missed activation. Could you make your question more clear next time? Thanks. 

For layers of discriminator, I don’t have BN either. I don’t have any further suggestion.

**Comment by JiahuiYu** [comment]
The final layer of G is not gated as I already before. It is a simple convolution with tf.tanh as activation. 

Usually the blur means you discriminator May have bug. GAN makes results more realistic.

**Comment by huangqianfirst** [comment]
I see . Thank you so much for your patience.

**Comment by JiahuiYu** [comment]
No problem. I do hope that I could help. Thanks for your interest in our work.

**Comment by an1018** [comment]
@huangqianfirst I meet the same problem. Have you solved it?Looking forword your reply!

**Comment by huangqianfirst** [comment]
Not yet. But i still doing on it.

**Comment by an1018** [comment]
@huangqianfirst Is there any improvement through increasing iters?And what is your maximum number of iters?

**Comment by huangqianfirst** [comment]
I'm in training. I'm not sure whether more iters can improve my result.

**Comment by JiahuiYu** [comment]
@huangqianfirst @an1018 I wonder if both of you have successfully trained the code of deep fill v1 with rectangular masks? How many days are required to train a good performance generator on your GPU machines? The reproduced results of both of you are much worse than my trained model in these examples.

Are you students? Chinese? Maybe I can create a group for both of you to discuss what is the potential problem of your implementation.

**Comment by huangqianfirst** [comment]
I'm Chinese. My wechat is hq18745018992.

**Comment by an1018** [comment]
@JiahuiYu I have sent my WeChat to your email

**Comment by Haoyanlong** [comment]
@JiahuiYu @huangqianfirst @an1018 ,I want to join the group, my WeChat is 13720633429,Could you let me join the group?Thank you very much.I have trained the model on my dataset by PyTorch.My visualization of tensorboard as flows(masked image, coarse output, refine output, ori scene image, last output.):
![image](https://user-images.githubusercontent.com/23288056/76858560-a28ec180-6892-11ea-9ce5-dbb81ceb139c.png)
